### ISSUE TITLE
There are messages of error in file 'sql-errors.log ' (III)

### ISSUE THREAD

**Issue by ddurieux** [fusioninventory/fusioninventory-for-glpi#1547]
---

Author Name: **Tomás Abad** (Tomás Abad)
Original Redmine Issue: 2768, http://forge.fusioninventory.org/issues/2768
Original Date: 2014-10-28

---

## The environment

* GLPI 0.84.8
* FusionInventory for GLPI 0.84+3.4
* FusionInventory Agent 2.3.10.1 (99% of them)

## The error message

```
2014-10-28 08:44:29
*** MySQL query error: 
***
SQL: UPDATE `glpi_computers_softwareversions`
                         SET `softwareversions_id` = \'96241\'
                         WHERE `id` = 2773068
Error: Duplicate entry '7337-96241' for key 'unicity'
Backtrace :
/usr/local/share/glpi/inc/transfer.class.php&nbsp;:1442         DBmysql->query()
/usr/local/share/glpi/inc/transfer.class.php&nbsp;:1053         Transfer->transferComputerSoftwares()
/usr/local/share/glpi/inc/transfer.class.php&nbsp;:183          Transfer->transferItem()
/usr/local/share/glpi/plugins/fusioninventory/inc/inventorycomputerinventory.class.php&nbsp;:462                Transfer->moveItems()
/usr/local/share/glpi/plugins/fusioninventory/inc/inventoryruleimport.class.php&nbsp;:664               PluginFusioninventoryInventoryComputerInventory->rulepassed()
/usr/local/share/glpi/inc/rule.class.php&nbsp;:997              PluginFusioninventoryInventoryRuleImport->executeActions()
/usr/local/share/glpi/inc/rulecollection.class.php&nbsp;:695            Rule->process()
/usr/local/share/glpi/plugins/fusioninventory/inc/inventorycomputerinventory.class.php&nbsp;:350                RuleCollection->processAllRules()
/usr/local/share/glpi/plugins/fusioninventory/inc/inventorycomputerinventory.class.php&nbsp;:104                PluginFusioninventoryInventoryComputerInventory->sendCriteria()
/usr/local/share/glpi/plugins/fusioninventory/inc/communication.class.php&nbsp;:222             PluginFusioninventoryInventoryComputerInventory->import()
/usr/local/share/glpi/plugins/fusioninventory/inc/communication.class.php&nbsp;:438             PluginFusioninventoryCommunication->import()
/usr/local/share/glpi/plugins/fusioninventory/front/communication.php&nbsp;:85          PluginFusioninventoryCommunication->handleOCSCommunication()
/usr/local/share/glpi/plugins/fusioninventory/index.php&nbsp;:50                include_once()
/usr/local/share/glpi/plugins/fusioninventory/index.php

2014-10-28 17:52:17
*** MySQL query error: 
***
SQL: UPDATE `glpi_computers_softwareversions`
                         SET `softwareversions_id` = \'88126\'
                         WHERE `id` = 2787926
Error: Duplicate entry '7653-88126' for key 'unicity'
Backtrace :
/usr/local/share/glpi/inc/transfer.class.php&nbsp;:1442         DBmysql->query()
/usr/local/share/glpi/inc/transfer.class.php&nbsp;:1053         Transfer->transferComputerSoftwares()
/usr/local/share/glpi/inc/transfer.class.php&nbsp;:183          Transfer->transferItem()
/usr/local/share/glpi/plugins/fusioninventory/inc/inventorycomputerinventory.class.php&nbsp;:462                Transfer->moveItems()
/usr/local/share/glpi/plugins/fusioninventory/inc/inventoryruleimport.class.php&nbsp;:664               PluginFusioninventoryInventoryComputerInventory->rulepassed()
/usr/local/share/glpi/inc/rule.class.php&nbsp;:997              PluginFusioninventoryInventoryRuleImport->executeActions()
/usr/local/share/glpi/inc/rulecollection.class.php&nbsp;:695            Rule->process()
/usr/local/share/glpi/plugins/fusioninventory/inc/inventorycomputerinventory.class.php&nbsp;:350                RuleCollection->processAllRules()
/usr/local/share/glpi/plugins/fusioninventory/inc/inventorycomputerinventory.class.php&nbsp;:104                PluginFusioninventoryInventoryComputerInventory->sendCriteria()
/usr/local/share/glpi/plugins/fusioninventory/inc/communication.class.php&nbsp;:222             PluginFusioninventoryInventoryComputerInventory->import()
/usr/local/share/glpi/plugins/fusioninventory/inc/communication.class.php&nbsp;:438             PluginFusioninventoryCommunication->import()
/usr/local/share/glpi/plugins/fusioninventory/front/communication.php&nbsp;:85          PluginFusioninventoryCommunication->handleOCSCommunication()
/usr/local/share/glpi/plugins/fusioninventory/index.php&nbsp;:50                include_once()
/usr/local/share/glpi/plugins/fusioninventory/index.php


```

**Comment by ddurieux** [comment]
---

Original Redmine Comment
Author Name: **Tomás Abad** (Tomás Abad)
Original Date: 2014-11-04T12:08:11Z

---

And again with FusionInventory for GLPI 0.84+3.5.

```
2014-11-04 09:26:15
*** MySQL query error:
***
SQL: UPDATE `glpi_computers_softwareversions`
                         SET `softwareversions_id` = \'109650\'
                         WHERE `id` = 3002036
Error: Duplicate entry '7337-109650' for key 'unicity'
Backtrace :
/usr/local/share/glpi/inc/transfer.class.php&nbsp;:1442         DBmysql->query()
/usr/local/share/glpi/inc/transfer.class.php&nbsp;:1053         Transfer->transferComputerSoftwares()
/usr/local/share/glpi/inc/transfer.class.php&nbsp;:183          Transfer->transferItem()
/usr/local/share/glpi/plugins/fusioninventory/inc/inventorycomputerinventory.class.php&nbsp;:462                Transfer->moveItems()
/usr/local/share/glpi/plugins/fusioninventory/inc/inventoryruleimport.class.php&nbsp;:664               PluginFusioninventoryInventoryComputerInventory->rulepassed()
/usr/local/share/glpi/inc/rule.class.php&nbsp;:997              PluginFusioninventoryInventoryRuleImport->executeActions()
/usr/local/share/glpi/inc/rulecollection.class.php&nbsp;:695            Rule->process()
/usr/local/share/glpi/plugins/fusioninventory/inc/inventorycomputerinventory.class.php&nbsp;:350                RuleCollection->processAllRules()
/usr/local/share/glpi/plugins/fusioninventory/inc/inventorycomputerinventory.class.php&nbsp;:104                PluginFusioninventoryInventoryComputerInventory->sendCriteria()
/usr/local/share/glpi/plugins/fusioninventory/inc/communication.class.php&nbsp;:222             PluginFusioninventoryInventoryComputerInventory->import()
/usr/local/share/glpi/plugins/fusioninventory/inc/communication.class.php&nbsp;:438             PluginFusioninventoryCommunication->import()
/usr/local/share/glpi/plugins/fusioninventory/front/communication.php&nbsp;:85          PluginFusioninventoryCommunication->handleOCSCommunication()
/usr/local/share/glpi/plugins/fusioninventory/index.php&nbsp;:50                include_once()
/usr/local/share/glpi/plugins/fusioninventory/index.php


```

**Comment by ddurieux** [comment]
---

Original Redmine Comment
Author Name: **Tomás Abad** (Tomás Abad)
Original Date: 2014-11-08T01:20:23Z

---

And another one...

```
2014-11-06 09:30:27
*** MySQL query error: 
***
SQL: UPDATE `glpi_computers_softwareversions`
                         SET `softwareversions_id` = \'107107\'
                         WHERE `id` = 3002077
Error: Duplicate entry '7337-107107' for key 'unicity'
Backtrace :
/usr/local/share/glpi/inc/transfer.class.php&nbsp;:1442         DBmysql->query()
/usr/local/share/glpi/inc/transfer.class.php&nbsp;:1053         Transfer->transferComputerSoftwares()
/usr/local/share/glpi/inc/transfer.class.php&nbsp;:183          Transfer->transferItem()
/usr/local/share/glpi/plugins/fusioninventory/inc/inventorycomputerinventory.class.php&nbsp;:462                Transfer->moveItems()
/usr/local/share/glpi/plugins/fusioninventory/inc/inventoryruleimport.class.php&nbsp;:664               PluginFusioninventoryInventoryComputerInventory->rulepassed()
/usr/local/share/glpi/inc/rule.class.php&nbsp;:997              PluginFusioninventoryInventoryRuleImport->executeActions()
/usr/local/share/glpi/inc/rulecollection.class.php&nbsp;:695            Rule->process()
/usr/local/share/glpi/plugins/fusioninventory/inc/inventorycomputerinventory.class.php&nbsp;:350                RuleCollection->processAllRules()
/usr/local/share/glpi/plugins/fusioninventory/inc/inventorycomputerinventory.class.php&nbsp;:104                PluginFusioninventoryInventoryComputerInventory->sendCriteria()
/usr/local/share/glpi/plugins/fusioninventory/inc/communication.class.php&nbsp;:222             PluginFusioninventoryInventoryComputerInventory->import()
/usr/local/share/glpi/plugins/fusioninventory/inc/communication.class.php&nbsp;:438             PluginFusioninventoryCommunication->import()
/usr/local/share/glpi/plugins/fusioninventory/front/communication.php&nbsp;:85          PluginFusioninventoryCommunication->handleOCSCommunication()
/usr/local/share/glpi/plugins/fusioninventory/index.php&nbsp;:50                include_once()
/usr/local/share/glpi/plugins/fusioninventory/index.php


```

**Comment by SylvieCozic** [comment]
Hello,

I had the same issue (with OCSinventoryNG plugin though).
At the time i am writing, the only case I encountered was when I was transfering a computer from one entity to another one and this computer had 2 rows with the same software name but with 2 differents manufacturers.
For example "PDFCreator" from "PDFForge" and "PDFCreator" from "Frank Heindörfer, Philip Chinery"

The issue was with the function copySingleSoftware wich is called by transferComputerSoftwares().
This function only check the software name and not the manufacturer and return the first software found.
In the "PDFCreator" case it was trying to transfer the same "softwareversion" twice --> error.

SO ! In transfer.class.php, on line 1409, I changed the query to :

            $query = "SELECT *
                      FROM `glpi_softwares`
                      WHERE `entities_id` = ".$this->to."
                            AND `name` = '".addslashes($soft->fields['name'])."'
                            AND `manufacturers_id` = '".$soft->fields['manufacturers_id']."'";

I don't have the "Duplicate entry for key 'unicity'" error anymore so far.

Regards,

Sylvie
